### PR TITLE
Enable to use polygon and bitmap mask as reference mask for zero-shot learning

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes:
- [x] Enable polygon and bitmap mask as reference mask

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/model/visual_prompting.py [hunk 1]
```diff
@@ -923,7 +926,7 @@ def _customize_inputs(  # type: ignore[override]
             if self.training:
                 points: list[Prompt] = []
                 bboxes: list[Prompt] = []
-                for prompt, label in zip(prompts, labels):  # type: ignore[arg-type]
+                for prompt, label in zip(prompts, labels["prompts"]):  # type: ignore[arg-type]
```
Comment (sovrasov) [inline src/otx/core/model/visual_prompting.py]: Are polygons supported by OTX OV model? MAPI wrapper already supports them, so we can just collect & forward the polygons here.

File: src/otx/core/data/entity/visual_prompting.py [hunk 2]
```diff
@@ -137,38 +133,39 @@ class ZeroShotVisualPromptingDataEntity(OTXDataEntity):
...
+        polygons (list[list[dmPolygon]]): List of polygons.
+        prompts (list[list[tvBoundingBoxes | Points]]): List of prompts.
     """
 
-    masks: list[tv_tensors.Mask]
-    labels: list[LongTensor]
-    polygons: list[list[Polygon]]
-    prompts: list[list[tv_tensors.TVTensor]]
+    masks: list[tvMask]
+    labels: list[dict[Literal["prompts", "polygons", "masks"], LongTensor]]
```
Comment (sovrasov) [inline src/otx/core/data/entity/visual_prompting.py]: Isn't it time to introduce a dataclass instead of dict?

File: src/otx/core/data/dataset/visual_prompting.py [hunk 3]
```diff
@@ -76,24 +86,21 @@ def _get_item_impl(self, index: int) -> VisualPromptingDataEntity | None:
...
                     # skip very small region
                     continue
 
                 if torch.rand(1) < self.prob:
                     # get bbox
-                    bbox = tv_tensors.BoundingBoxes(
+                    bbox = tvBoundingBoxes(
                         annotation.get_bbox(),
-                        format=tv_tensors.BoundingBoxFormat.XYWH,
+                        format="xywh",
```
Comment (sovrasov) [inline src/otx/core/data/dataset/visual_prompting.py]: `BoundingBoxFormat.XYWH` seems to be more reliable than a literal

File: src/otx/core/data/dataset/visual_prompting.py [hunk 4]
```diff
@@ -231,22 +246,24 @@ def _get_item_impl(self, index: int) -> ZeroShotVisualPromptingDataEntity | None
...
+                gt_labels["masks"].append(annotation.label)
                 gt_masks.append(mask)
                 gt_polygons.append(annotation)
 
             # TODO(sungchul): for mask, bounding box, and point annotation
             elif isinstance(annotation, (dmBbox, dmMask, dmPoints)):
                 pass
 
-        assert len(gt_prompts) > 0, "#prompts must be greater than 0."  # noqa: S101
+        if len(gt_prompts) == 0:
```
Comment (eunwoosh) [inline src/otx/core/data/dataset/visual_prompting.py]:
```suggestion
        if not gt_prompts:
```
It's a minor but how about aligning it to PEP8? This is a reference https://peps.python.org/pep-0008/#programming-recommendations

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py [hunk 5]
```diff
@@ -745,7 +742,14 @@ def _customize_inputs(  # type: ignore[override]
...
             # learn
             forward_inputs.update(
-                {"processed_prompts": self._gather_prompts_with_labels(inputs.prompts, inputs.labels)},
+                {
+                    "processed_prompts": self._gather_prompts_with_labels(
+                        inputs.labels,
+                        inputs.prompts,
+                        inputs.polygons,
+                        inputs.masks,
+                    ),
```
Comment (eunwoosh) [inline src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: It seems that `_gather_prompts_with_labels` function reuiqres all attributes of `ZeroShotVisualPromptingDataEntity`. If then, how about passing `inputs` rather than each attirbutes?

File: src/otx/core/data/dataset/visual_prompting.py [hunk 6]
```diff
@@ -231,22 +246,24 @@ def _get_item_impl(self, index: int) -> ZeroShotVisualPromptingDataEntity | None
                     )
                     gt_prompts.append(point)
 
-                gt_labels.append(annotation.label)
+                gt_labels["prompts"].append(annotation.label)
+                gt_labels["polygons"].append(annotation.label)
+                gt_labels["masks"].append(annotation.label)
```
Comment (eunwoosh) [inline src/otx/core/data/dataset/visual_prompting.py:253]: It's just a question. If same list is assigned to prompts, polygons and masks, isn't it possible to keep single list?

File: src/otx/core/model/visual_prompting.py [hunk 7]
```diff
@@ -130,6 +130,9 @@ def _inference_step_for_zero_shot(
     if not isinstance(preds, ZeroShotVisualPromptingBatchPredEntity):
         raise TypeError(preds)
 
+    # filter labels using corresponding ground truth
+    inputs.labels = [label["masks"] if len(inputs.masks) > 0 else label["polygons"] for label in inputs.labels]
```
Comment (eunwoosh) [inline src/otx/core/model/visual_prompting.py]:
```suggestion
    inputs.labels = [label["masks"] if inputs.masks else label["polygons"] for label in inputs.labels]
```
Same as above

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left some comments. Please take a look.

File: src/otx/core/model/visual_prompting.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/core/model/visual_prompting.py]: I updated polygon support at L938-940 to use MAPI wrapper's polygon component.
https://github.com/openvinotoolkit/training_extensions/blob/58934910c3c08d3a7106c863539f28d179492ce3/src/otx/core/model/visual_prompting.py#L938-L940

File: src/otx/core/data/dataset/visual_prompting.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/core/data/dataset/visual_prompting.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/3769/commits/a425190b5050f81a5f10cfd3fd6432bfb3e43338

File: src/otx/core/data/entity/visual_prompting.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/core/data/entity/visual_prompting.py]: Updated ab08d68

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Updated https://github.com/openvinotoolkit/training_extensions/commit/ab08d6845bdca69cecb90482049e6e32d07aa4fe

File: src/otx/core/model/visual_prompting.py (hunk 7 above)
Comment (sungchul1) [reply src/otx/core/model/visual_prompting.py]: Updated https://github.com/openvinotoolkit/training_extensions/commit/ab08d6845bdca69cecb90482049e6e32d07aa4fe

File: src/otx/core/data/dataset/visual_prompting.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/core/data/dataset/visual_prompting.py]: Done 3f1333f

File: src/otx/core/data/dataset/visual_prompting.py (hunk 6 above)
Comment (sungchul1) [reply src/otx/core/data/dataset/visual_prompting.py:253]: The purpose of this implementation is to distinguish various inputs' labels.
Not implemented yet, but if datumaro boundingbox or point is only given wihout polygon or mask, labels of polygon and mask will not be set.
It is the implementation for zero shot learning that can get dynamic input types.

**Review by eunwoosh — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]
